huntsman/observatory.py

```diff
@@ -232,202 +233,67 @@ def autofocus_cameras(self, *args, **kwargs):
-                self.logger.debug("Too many flats, quitting")
-                break
-
-            self.logger.debug("Checking for saturation")
-            if is_saturated and exptimes[cam_name][-1].value < 2:
-                self.logger.debug(
-                    "Saturated with short exposure, waiting 30 seconds before next exposure")
-                max_num_exposures += 1
-                time.sleep(30)
-
-        # Add a bias exposure
-        for cam_name in camera_list:
-            exptimes[cam_name].append(0 * u.second)
-
-        # Record how many exposures we took
-        num_exposures = flat_obs.current_exp_num
-
-        # Take darks (just last ten)
-        for i in range(num_exposures):
-            self.logger.debug("Slewing to dark-field coords: {}".format(flat_obs.field))
-            self.mount.set_target_coordinates(flat_obs.field)
-            self.mount.slew_to_target()
-            self.status()
-
-            for cam_name in camera_list:
-
-                try:
-                    exptime = exptimes[cam_name][i]
-                except KeyError:
-                    break
-
-                camera = self.cameras[cam_name]
-
-                filename = "{}/flats/{}/{}/{}.{}".format(
-                    image_dir,
-                    camera.uid,
-                    flat_obs.seq_time,
-                    'dark_{:02d}'.format(flat_obs.current_exp_num),
-                    camera.file_extension)
-
-                # Take picture and wait for result
-                camera_event = camera.take_observation(
-                    flat_obs,
-                    fits_headers,
-                    filename=filename,
-                    exptime=exptime,
-                    dark=True
-                )
-
-                camera_events[cam_name] = {
-                    'event': camera_event,
-                    'filename': filename,
-                }
-
-            # Will block here until done exposing on all cameras
-            while not all([info['event'].is_set() for info in camera_events.values()]):
-                self.logger.debug('Waiting for dark-field image')
-                time.sleep(1)
-
-    def take_morning_flats(*args, **kwargs):
-        '''
-
-        '''
-        raise NotImplementedError('Morning flats not implemented yet.')
+        if camera_names is None:
+            cameras_all = self.cameras
+        else:
+            cameras_all = {c: self.cameras[c] for c in camera_names}
+
+        # Obtain the filter order
+        filter_order = self.config['filterwheel']['flat_field_order']
+        if self.past_midnight:  # If it's the morning, order is reversed
+            filter_order = filter_order[::-1]
+
+        exptimes_dark = {c: [0 * u.second] for c in cameras_all.keys()}  # Add a bias exposure
+        for filter_name in filter_order:
+
+            # Get a dict of cameras that have this filter
+            cameras = {}
```

Comment:
Suggest renaming this variable.
```suggestion
            filter_cameras = {}
```